Sparse_and_dense.ipynb

```diff
@@ -0,0 +1,320 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 41,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "from scipy import sparse\n",
+    "import sys\n",
+    "import os\n",
+    "import time"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 89,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "np.random.seed(0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Sparse .npz (compressed) takes the least disk space regardless of density\n",
+    "Uncompressed sparse matrices take more disk space when the matrices are very dense"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 40,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CCpGeAeGaD (1391, 137)\n",
+      "density: 97.8 %\n",
+      "dense: 1.525 MB\n",
+      "sparse: 1.347 MB\n",
+      "sparse_uncompressed: 2.238 MB\n",
+      "\n",
+      "DaGpBP (137, 11381)\n",
+      "density: 21.0 %\n",
+      "dense: 12.47 MB\n",
+      "sparse: 2.529 MB\n",
+      "sparse_uncompressed: 3.982 MB\n",
+      "\n",
+      "MFpGdCcSE (2884, 5734)\n",
+      "density: 14.1 %\n",
+      "dense: 132.3 MB\n",
+      "sparse: 17.79 MB\n",
+      "sparse_uncompressed: 27.99 MB\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "for i in ['CCpGeAeGaD', 'DaGpBP', 'MFpGdCcSE']:\n",
+    "    mat = np.load(f'data/{i}.npy')\n",
+    "    print(i, mat.shape)\n",
+    "    \n",
+    "    density = 100 * (mat != 0).sum() / np.prod(mat.shape)\n",
+    "    print(f'density: {density :.3} %')\n",
+    "\n",
+    "    sparse_mat = sparse.csc_matrix(mat, copy=True)\n",
+    "\n",
+    "    sparse.save_npz(f'data/{i}_sparse', sparse_mat)\n",
+    "    sparse.save_npz(f'data/{i}_sparse_un', sparse_mat, compressed=False)\n",
+    "    \n",
+    "    dense_size = os.path.getsize(f'data/{i}.npy') / 1000000\n",
+    "    sparse_size = os.path.getsize(f'data/{i}_sparse.npz') / 1000000\n",
+    "    sparse_un_size = os.path.getsize(f'data/{i}_sparse_un.npz') / 1000000\n",
+    "    print(f'dense: {dense_size :.4} MB\\nsparse: {sparse_size :.4} MB\\nsparse_uncompressed: {sparse_un_size :.4} MB\\n')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Memory-map reduces matrix-vector multiplication time for lower-density matrices\n",
+    "\n",
+    "Putting a normal load above a memmap load decreases memmap load time. However, loading memmap first does not speed up normal load times. Note, this only appears to be the case for very dense matrices. For best comparisons, we should always load a memmap matrix first in the sequence."
```

Comment:
It looks to me like the first time you call `np.load` with `mmap_mode='r'`, memory-mapped reading is slower. However, subsequent repeated reads of the same matrix in memory-mapped mode take similar time. This makes me think the slowness of the initial load has to do with importing the memory-map machinery and it's actually subsequent reads that we care about.